### DISCUSSION TITLE
How can I contribute for a rule to get stable status?

### DISCUSSION THREAD

**Discussion by R-Peleg** [astral-sh/ruff#14331] [Q&A]
There is a particular rule (E303) that we really want to have in stable mode, and willing to dedicate some time and effort to make it happen. Is there a way to contribute to this specific effort?

I understand that every rules has to be in Preview mode for some time, since confidence about stability takes time and some things cannot be rushed. I found the discussion at https://github.com/astral-sh/ruff/discussions/6998 describing a rule will remain in Preview for at least one minor release, but seems to have some case-by-case guidelines for whether or not it happen on the next release.

I didn't see any open issues with that particular rule, so we can't try to fix them. Will running this rule on large open-source codebases and reporting the results help? Anything else we can do?

**Comment by MichaReiser** [comment]
Thanks for reaching out. The E rules are a bit special.

We've been hesitant to stabilize the `E3*` rules because they very closely overlap with the formatter and often implement the exact same logic. We also have to be extra careful to review every formatter or rule change so that they remain compatible, which is a painful process, and we missed multiple incompatibilities in the past. That's why we aren't 100% convinced they should be part of Ruff, which is also why we held back with stabilizing and adding similar rules.

**Reply by kaddkaka** [reply]
Rules should stay compatible with what? Other rules?

I posted my argument on why I think these rules should stabilized regardless if they are in the formatter or not.

https://github.com/astral-sh/ruff/issues/2402#issuecomment-2575283089
